lib/core/request.js

```diff
@@ -206,41 +247,63 @@ class Request {
       channels.bodySent.publish({ request: this })
     }
 
-    if (this[kHandler].onRequestSent) {
-      try {
-        return this[kHandler].onRequestSent()
-      } catch (err) {
-        this.abort(err)
-      }
+    try {
+      this[kHandler].onRequestEnd?.()
+      this[kHandler].onRequestSent?.()
+    } catch (err) {
+      this.abort(err)
     }
   }
 
   onConnect (abort) {
     assert(!this.aborted)
     assert(!this.completed)
+    assert(!this.controller)
 
     if (this.error) {
       abort(this.error)
     } else {
       this.abort = abort
-      return this[kHandler].onConnect(abort)
+      return this[kHandler].onConnect?.(abort)
     }
   }
 
-  onResponseStarted () {
-    return this[kHandler].onResponseStarted?.()
+  onResponseStarted (resume) {
+    assert(!this.aborted)
+    assert(!this.completed)
+
+    this.controller = new Controller(this.abort, resume)
+    try {
+      this[kHandler].onResponseStart?.(this.controller)
+
+      if (this[kHandler].onResponseStarted?.() === false) {
+        this.controller.pause()
+      }
+
+      return !this.controller.paused
+    } catch (err) {
+      this.abort(err)
+    }
   }
 
-  onHeaders (statusCode, headers, resume, statusText) {
+  onHeaders (headers, statusCode, statusText) {
     assert(!this.aborted)
     assert(!this.completed)
+    assert(this.controller)
 
     if (channels.headers.hasSubscribers) {
       channels.headers.publish({ request: this, response: { statusCode, headers, statusText } })
     }
 
     try {
-      return this[kHandler].onHeaders(statusCode, headers, resume, statusText)
+      this[kHandler].onResponseHeaders?.(parseHeaders(headers), statusCode, statusText)
```

Comment:
Here we are no longer paso g raw headers. We parse them immediately. As far as I can tell this is what most users do and it's unfortunate to force middleware to work on raw headers.

Reply:
@ShogunPanda @metcoder95 @mcollina any thoughts here? Is this a mistake (i.e. performance blocker) or just being pragmatic?

Reply:
The latter. Let's be pragmatic.

Reply:
+1 on pragmatism